cloud/scope/powervs_cluster.go

```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+		return nil, err
+	}
+	lbState := infrav1beta2.VPCLoadBalancerState(*loadBalancer.ProvisioningStatus)
+	return &infrav1beta2.VPCLoadBalancerStatus{
+		ID:                loadBalancer.ID,
+		State:             lbState,
+		Hostname:          loadBalancer.Hostname,
+		ControllerCreated: pointer.Bool(true),
+	}, nil
+}
+
+// COSInstance returns the COS instance reference.
+func (s *PowerVSClusterScope) COSInstance() *infrav1beta2.CosInstance {
+	return s.IBMPowerVSCluster.Spec.CosInstance
+}
+
+// ReconcileCOSInstance reconcile COS bucket.
+func (s *PowerVSClusterScope) ReconcileCOSInstance() error {
+	if s.COSInstance() == nil || s.COSInstance().Name == "" {
+		return nil
+	}
+
+	// check COS service instance exist in cloud
+	cosServiceInstanceStatus, err := s.checkCOSServiceInstance()
+	if err != nil {
+		s.Error(err, "error checking cos service instance")
+		return err
+	}
+	if cosServiceInstanceStatus != nil {
+		s.SetStatus(COSInstance, infrav1beta2.ResourceReference{ID: cosServiceInstanceStatus.GUID, ControllerCreated: pointer.Bool(false)})
+	} else {
+		// create COS service instance
+		cosServiceInstanceStatus, err = s.createCOSServiceInstance()
+		if err != nil {
+			s.Error(err, "error creating cos service instance")
+			return err
+		}
+		s.SetStatus(COSInstance, infrav1beta2.ResourceReference{ID: cosServiceInstanceStatus.GUID, ControllerCreated: pointer.Bool(true)})
+	}
+
+	props, err := authenticator.GetProperties()
+	if err != nil {
+		s.Error(err, "error while fetching service properties")
+		return err
+	}
+
+	apiKey := props["APIKEY"]
+	if apiKey == "" {
+		fmt.Printf("ibmcloud api key is not provided, set %s environmental variable", "IBMCLOUD_API_KEY")
+	}
+
+	cosClient, err := cos.NewService(cos.ServiceOptions{}, s.IBMPowerVSCluster.Spec.CosInstance.BucketRegion, apiKey, *cosServiceInstanceStatus.GUID)
+	if err != nil {
+		s.Error(err, "error creating cosClient")
+		return fmt.Errorf("failed to create cos client: %w", err)
+	}
+	s.COSClient = cosClient
+
+	// check bucket exist in service instance
+	if exist, err := s.checkCOSBucket(); exist {
+		return nil
+	} else if err != nil {
+		s.Error(err, "error checking cos bucket")
+		return err
+	}
+
+	// create bucket in service instance
+	if err := s.createCOSBucket(); err != nil {
+		s.Error(err, "error creating cos bucket")
+		return err
+	}
+	return nil
+}
+
+func (s *PowerVSClusterScope) checkCOSBucket() (bool, error) {
+	if _, err := s.COSClient.GetBucketByName(s.COSInstance().BucketName); err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case s3.ErrCodeNoSuchBucket, "Forbidden", "NotFound":
+				// If the bucket doesn't exist that's ok, we'll try to create it
+				return false, nil
+			default:
+				return false, err
+			}
+		} else {
+			return false, err
+		}
+	}
+	return true, nil
+}
+
+func (s *PowerVSClusterScope) createCOSBucket() error {
+	input := &s3.CreateBucketInput{
+		Bucket: pointer.String(s.COSInstance().BucketName),
```

Comment:
There's a scenario where when we retry bucket creation immediately with the same name,  it throws an error saying bucket exists even though the service instance and bucket get deleted successfully. Maybe we can handle by adding a random string to the bucket name?
(check out `util.RandomString()` from `sigs.k8s.io/cluster-api/util`)